subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCacheAccess.java

```diff
@@ -499,4 +511,80 @@ Thread getOwner() {
+        }
+
+        public PersistentIndexedCacheParameters getParameters() {
+            return parameters;
+        }
+
+        public void assertCompatibleCacheParameters(PersistentIndexedCacheParameters parameters) {
+            List<String> faultMessages = new ArrayList<String>();
+
+            assertCacheNameMatch(faultMessages, parameters.getCacheName());
+            assertCompatibleKeySerializer(faultMessages, parameters.getKeySerializer());
+            assertCompatibleValueSerializer(faultMessages, parameters.getValueSerializer());
+            assertCompatibleCacheDecorator(faultMessages, parameters.getCacheDecorator());
+
+            if (!faultMessages.isEmpty()) {
+                String lineSeparator = SystemProperties.getInstance().getLineSeparator();
+                String faultMessage = StringUtils.join(faultMessages, lineSeparator);
+                throw new InvalidCacheReuseException(
+                    "The cache couldn't be reuse because of the following mismatch:" + lineSeparator + faultMessage);
+            }
+        }
+
+        private void assertCacheNameMatch(Collection<String> faultMessages, String cacheName) {
+            if (!Objects.equal(cacheName, parameters.getCacheName())) {
+                faultMessages.add(
+                    String.format(" * Requested cache name (%s) doesn't match current cache name (%s)", cacheName,
+                        parameters.getCacheName()));
+            }
+        }
+
+        private void assertCompatibleKeySerializer(Collection<String> faultMessages, Serializer keySerializer) {
+            if (!Objects.equal(keySerializer.getClass(), parameters.getKeySerializer().getClass())) {
+                faultMessages.add(
+                    String.format(" * Requested key serializer type (%s) doesn't match current cache type (%s)",
+                        keySerializer.getClass().getCanonicalName(),
+                        parameters.getKeySerializer().getClass().getCanonicalName()));
+            }
+        }
+
+        private void assertCompatibleValueSerializer(Collection<String> faultMessages, Serializer valueSerializer) {
+            if (!Objects.equal(valueSerializer.getClass(), parameters.getValueSerializer().getClass())) {
+                faultMessages.add(
+                    String.format(" * Requested value serializer type (%s) doesn't match current cache type (%s)",
+                        valueSerializer.getClass().getCanonicalName(),
+                        parameters.getValueSerializer().getClass().getCanonicalName()));
+            }
+        }
+
+        private void assertCompatibleCacheDecorator(Collection<String> faultMessages, CacheDecorator cacheDecorator) {
+            String requestClassName = ClassUtils.getShortCanonicalName(cacheDecorator, "null");
+            String currentClassName = ClassUtils.getShortCanonicalName(parameters.getCacheDecorator(), "null");
+            if (!Objects.equal(requestClassName, currentClassName)) {
```

Comment:
why is this testing class names vs class instances like the serializers?  If you're worried about null, I think `Objects.equal()` would handle that if one or the other cacheDecorator is null.